dupedetection/client.go

```diff
@@ -0,0 +1,127 @@
+package dupedetection
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/pastelnetwork/gonode/common/errors"
+	"github.com/pastelnetwork/gonode/common/log"
+	"github.com/pastelnetwork/gonode/common/random"
+)
+
+type client struct {
+	config *Config
+}
+
+// Generate implements dupedection.Client.Generate
+func (client *client) Generate(ctx context.Context, path string) (*DupeDetection, error) {
+
+	// Copy image file to dupe detection service input directory
+	outputBase, outputPath, err := client.copyImageToInputDir(path)
+	if err != nil {
+		return nil, errors.Errorf("failed to copy image to dupe detection input directory: %w", err)
+	}
+
+	// Collect result from dupe detection output directory
+	result, err := client.collectOutput(ctx, outputBase, outputPath)
+	if err != nil {
+		err = errors.Errorf("failed to collect output from dupe detection service: %w", err)
+	}
+
+	return result, err
+}
+
+func (client *client) copyImageToInputDir(inputPath string) (string, string, error) {
+	input, err := ioutil.ReadFile(inputPath)
+	if err != nil {
+		return "", "", err
+	}
+
+	ext := filepath.Ext(inputPath)
+
+	// Generate random name
+	inputName, err := random.String(10, random.Base62Chars)
+	if err != nil {
+		return "", "", err
+	}
+
+	inputPath = filepath.Join(client.config.InputDir, inputName+ext)
+
+	err = ioutil.WriteFile(inputPath, input, os.ModePerm)
+	if err != nil {
+		return "", "", err
+	}
+
+	outputBase := inputName + ".json"
+	outputPath := filepath.Join(client.config.OutputDir, outputBase)
+	return outputBase, outputPath, nil
+}
+
+func (client *client) collectOutput(ctx context.Context, baseName, path string) (*DupeDetection, error) {
+	// Waiting for JSON output from dupe detection service
+	ticker := time.NewTicker(1 * time.Second)
+	for {
+		select {
+		case <-ctx.Done():
+			ticker.Stop()
+			return nil, errors.New("Timeout")
+		case <-ticker.C:
+			files, err := ioutil.ReadDir(client.config.OutputDir)
+			if err != nil {
+				return nil, errors.Errorf("failed to scan dupe detection output directory: %v", err)
+			}
+			for _, file := range files {
+				if file.Name() == baseName {
+					result, err := parseOutput(path)
+
+					if err != nil {
+						log.WithContext(ctx).WithError(err).Debug("Failed to collect output from dupe detection service")
+						continue
+					}
+
+					return result, nil
+				}
+			}
+		}
+	}
+}
+
+func parseOutput(path string) (*DupeDetection, error) {
```

Comment:
Should have a timeout here because if dd-service is crashed/restarted and doesn't pick up left-over files then this thread will leak/hang until caller abort.

Reply:
I also considered this point. But not sure about how long it takes from dd-service side, so I decided to let the caller stop it via the context of the caller.

Reply:
okay, I added a timeout.